src/sigma/remint.cpp

```diff
@@ -0,0 +1,111 @@
+#include "remint.h"
+#include "zerocoin.h"
+#include "libzerocoin/Commitment.h"
+
+namespace sigma {
+
+CoinRemintToV3::CoinRemintToV3(unsigned mintVersion, unsigned denomination, unsigned groupId, Bignum serial, Bignum randomness,
+                    uint256 originalMintBlockHash, const std::vector<unsigned char> &ecdsaPrivateKey) {
+    coinRemintVersion = 1;
+    coinMintVersion = mintVersion;
+    coinDenomination = denomination;
+    coinGroupId = groupId;
+    coinSerial = serial;
+    coinRandomness = randomness;
+    coinPublicValue = CalculatePublicValue();
+
+    if (coinMintVersion > ZEROCOIN_TX_VERSION_1) {
+        this->ecdsaPrivateKey = ecdsaPrivateKey;
+        if (coinSerial.bitSize() > 160)
+            throw ZerocoinException("Invalid zerocoin mint");
+
+        secp256k1_pubkey pubkey;
+
+		if (!secp256k1_ec_pubkey_create(libzerocoin::ctx, &pubkey, ecdsaPrivateKey.data()))
+			throw ZerocoinException("Invalid secret key");
+
+        size_t len = 33;
+        ecdsaPublicKey = std::vector<unsigned char>(len, 0);
+        secp256k1_ec_pubkey_serialize(libzerocoin::ctx, ecdsaPublicKey.data(), &len, &pubkey, SECP256K1_EC_COMPRESSED);
+    }
+}
+
+void CoinRemintToV3::SignTransaction(const libzerocoin::SpendMetaData &metadata) {
+    if (ecdsaPrivateKey.empty())
+        throw ZerocoinException("Invalid remint transaction");
+
+    secp256k1_ecdsa_signature sig;
+    uint256 metahash = GetMetadataHash(metadata);
+	
+    if (secp256k1_ecdsa_sign(libzerocoin::ctx, &sig, metahash.begin(), ecdsaPrivateKey.data(), NULL, NULL) != 1)
+        throw ZerocoinException("Cannot sign remint transaction");
+
+    ecdsaSignature = vector<unsigned char>(64, 0);
+    secp256k1_ecdsa_signature_serialize_compact(libzerocoin::ctx, ecdsaSignature.data(), &sig);
+}
+
+void CoinRemintToV3::ClearSignature() {
+    ecdsaSignature.clear();
+}
+
+Bignum CoinRemintToV3::CalculatePublicValue() const {
+    // Parameters of coin committment group are always the same, we can take it from ZCParamsV2
+    const libzerocoin::IntegerGroupParams &commGroup = ZCParamsV2->coinCommitmentGroup;
+    // calculate g^serial * h^randomness (mod modulus)
+    return (commGroup.g.pow_mod(coinSerial, commGroup.modulus).mul_mod(commGroup.h.pow_mod(coinRandomness, commGroup.modulus), commGroup.modulus));
+}
+
+
+uint256 CoinRemintToV3::GetMetadataHash(const libzerocoin::SpendMetaData &metadata) const {
+    CHashWriter h(0,0);
+	h << metadata << coinSerial << coinPublicValue;
+    return h.GetHash();
+}
+
+bool CoinRemintToV3::Verify(const libzerocoin::SpendMetaData &metadata) const {
+    // check integer constants to be within allowed ranges
+
+    if (coinMintVersion != ZEROCOIN_TX_VERSION_1 && coinMintVersion != ZEROCOIN_TX_VERSION_2)
```

Comment:
There was a version 1_5 no?

Reply:
No, there wasn't. V1.5 applied to spend only